ts/test/types/message/initializeAttachmentMetadata_test.ts

```diff
@@ -0,0 +1,50 @@
+import 'mocha';
+import { assert } from 'chai';
+
+import * as Message from '../../../../ts/types/message/initializeAttachmentMetadata';
+import { IncomingMessage } from '../../../../ts/types/Message';
+import { MIMEType } from '../../../../ts/types/MIME';
+// @ts-ignore
+import { stringToArrayBuffer } from '../../../../js/modules/string_to_array_buffer';
+
+
+describe('Message', () => {
+  describe('initializeAttachmentMetadata', () => {
+    it('should handle visual media attachments', async () => {
+      const input: IncomingMessage = {
+        type: 'incoming',
+        conversationId: 'foo',
+        id: '11111111-1111-1111-1111-111111111111',
+        timestamp: 1523317140899,
+        received_at: 1523317140899,
+        sent_at: 1523317140800,
+        attachments: [{
+          contentType: 'image/jpeg' as MIMEType,
+          data: stringToArrayBuffer('foo'),
+          fileName: 'foo.jpg',
+          size: 1111,
+        }],
+      };
+      const expected: IncomingMessage = {
+        type: 'incoming',
+        conversationId: 'foo',
+        id: '11111111-1111-1111-1111-111111111111',
+        timestamp: 1523317140899,
+        received_at: 1523317140899,
+        sent_at: 1523317140800,
+        attachments: [{
+          contentType: 'image/jpeg' as MIMEType,
+          data: stringToArrayBuffer('foo'),
+          fileName: 'foo.jpg',
+          size: 1111,
+        }],
+        hasAttachments: 1,
+        hasVisualMediaAttachments: 1,
+        hasFileAttachments: 0,
```

Comment:
Ah, I see the answer to a previous question here. If one of these three booleans is false, it gets the value zero. One thing to consider here is that all of these indices will include every message in the database. That will likely have an impact on performance of lookups with the index.

It might be more efficient to change it to `1` | 'no key' which would ensure that the index only had the target elements in it. Given that the number of messages sent without attachments is so much higher than the number with (for the average user), I would recommend we do it that way.

Reply:
I went back and forth on this. I imagine lookup performance shouldn’t be affected as the index can be first traversed based on `hasFileAttachments: 1 | 0` and then the timestamps. However, it will have an impact on disk size of the index but I am willing to make that trade off for the future when we want to query our data in a more flexible way, e.g. find all messages without attachments, etc. Basically, this approach emulates what we’d get for free using SQLite or similar.